hazelcast/src/main/java/com/hazelcast/query/impl/bitmap/BitmapAlgorithms.java

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.query.impl.bitmap;
+
+/**
+ * Provides algorithms crucial for set operations on ordered iterators provided
+ * by sparse bit sets.
+ */
+final class BitmapAlgorithms {
+
+    private BitmapAlgorithms() {
+    }
+
+    /**
+     * @return an iterator that represents a result of intersection of the given
+     * iterators.
+     */
+    public static AscendingLongIterator and(AscendingLongIterator[] iterators) {
+        return new AndIterator(iterators);
+    }
+
+    /**
+     * @return an iterator that represents a result of union over the given
+     * iterators.
+     */
+    public static AscendingLongIterator or(AscendingLongIterator[] iterators) {
+        return new OrIterator(iterators);
+    }
+
+    /**
+     * @return an iterator that represents a result of negation of the given
+     * iterator over the given universe (a set of known elements).
+     */
+    public static AscendingLongIterator not(AscendingLongIterator iterator, SparseArray<?> universe) {
+        return new NotIterator(iterator, universe);
+    }
+
+    private static final class AndIterator implements AscendingLongIterator {
+
+        // The idea: iteratively advance all iterators to the current
+        // maximum index among all iterators until all iterators are at the
+        // same index or the end is reached in at least one of the iterators.
+
+        private final AscendingLongIterator[] iterators;
+
+        private long index;
+
+        AndIterator(AscendingLongIterator[] iterators) {
+            this.iterators = iterators;
+            advance();
+        }
+
+        @Override
+        public long getIndex() {
+            return index;
+        }
+
+        @Override
+        public long advance() {
+            long current = index;
+
+            long currentMax = iterators[0].getIndex();
```

Comment:
It seems we can use a bit more advanced algorithm to find a match. We rely on the fact that all indexes (positions) returned by the iterators are ordered ascending. In addition we sort the iterators by the current index (e.g. using priority queue) and then try to match the first iterator with the last one so that try to perform `firstIterator.advanceAtLeastTo(lastIterator.index)` skipping matching iterators in between. OK, if we find a match (index of the first iterator == index of the last iterator), otherwise the firstIterator becomes the last one (because it advanced beyond the current last one) and we repeat the process of matching the first and last iterators.

Reply:
done https://github.com/hazelcast/hazelcast/pull/16162/commits/4293ee1d0b9cd8bf054fbca7a6618da931943dda, many thanks for the idea